### DISCUSSION TITLE
Feature proposal for openjscad website url scheme

### DISCUSSION THREAD

**Discussion by hrgdavor** [jscad/OpenJSCAD.org#837] [Ideas]
Before creating an issue to implement I would like to propose an url scheme that would be useful in the long run

The basis is ofc the website that already has defined at least one parameter
 - uri=uri-to-a-script

I would like to propose pushing view options to the url as well
 - grid=0/1
 - zoomToFit=0/1
 - autoRotate=0/1
 - upload=0/1 (to hide or show project upload and enable/disable drag/drop of files)
 - download=0/1
 - editor=0/1
 - parameters=0/1

and atop of the url use # hash to store model parameter values in url format

this way it would be easy to embed openjscad as an iframe, and also supply parameters to the model
actually if web would react to hash change, one could emebed model in own page as iframe,
then supply parameters via own custom form into the iframe by changing the url

example url would be like this:

https://openjscad.xyz/?uri=https://mypage.com/modesl/something.js&grid=0&autoRotate=1&zoomToFit=1#radius=12&numberOfHoles=5

```
https://openjscad.xyz/

?uri=https://mypage.com/modesl/something.js
&grid=0
&autoRotate=1
&zoomToFit=1

#radius=12
&numberOfHoles=5
```

this would allow sharing of model url together with selected options and giving people simple embed iframe to paste in their website like with google maps or similar

In addition to url, few useful postMessage events could be allowed for using the embeded openjscad iframe

**loadScript** 
- parent website could supply script to the iframe allowing iframe to have script changed 
- this has potential security issues but looks like worker is isolated and scripts running there  should not be able to steal cookies or access tokens. (this will need to be confirmed manually, just to be sure)

**Comment by glennerichall** [comment]
I was just thinking that too. I was modeling a fan holder for my ender 3 and after desgining it in fusion 360, I came to the conclusion that it was not easily sharable, since I had to export multiple STL files for every possible printer configurations. So I tried OpenSCAD and finaly JScad. After all, why re-event a new language when there are plenty out there.

I was then able to script conditions to effectively select the parameters of the print. But then, non programming users must know how to fetch code from git (I shared it on git, see link below) and import it in openjscad.xyz. So I thought that maybe, like all online HTML editors, there must be an effective way to share designs online.

Your suggestion to use an external url I think is the way to go. But I would suggest maybe considering fetching data from a git repo as well. This should be easyer for developpers to share their project through openjscad.xyz. It would then be easy to provide a link directly on thingiverse for example to let users customize the print.

[Ender 3 Easy Removable Fan](https://github.com/glennerichall/Ender3-Removable-Fan)

**Reply by hrgdavor** [reply]
@glennerichall 
> But I would suggest maybe considering fetching data from a git repo as well.

It should not be a big problem to build it into a single file, that can then more easily be linked on openjscad.xyz

**Comment by hrgdavor** [comment]
There is a possibility to implement loading models from github, as they have the CORS headers to allow fetching directly from jscad without proxy
